src/rpc/core_rpc_server.cpp

```diff
@@ -2078,10 +2078,23 @@ namespace cryptonote
   //------------------------------------------------------------------------------------------------------------------------------
   bool core_rpc_server::on_get_output_distribution(const COMMAND_RPC_GET_OUTPUT_DISTRIBUTION::request& req, COMMAND_RPC_GET_OUTPUT_DISTRIBUTION::response& res, epee::json_rpc::error& error_resp)
   {
+    PERF_TIMER(on_get_output_distribution);
     try
     {
       for (uint64_t amount: req.amounts)
       {
+        static boost::mutex mutex;
+        boost::unique_lock<boost::mutex> lock(mutex);
+        static std::vector<uint64_t> cached_distribution;
+        static uint64_t cached_from = 0, cached_to = 0, cached_start_height = 0, cached_base = 0;
+        static bool cached = false;
+
+        if (cached && amount == 0 && cached_from == req.from_height && cached_to == req.to_height)
```

Comment:
I'm not sure how much this helps the wallet right now, but its possible to do a range selection (`cached_from <= req.from_height && cached_to >= req.to_height`), then do a range copy:
```c++
using vcopy = boost::adaptors::copy;
const std::size_t copy_start = req.from_height - cached_from;
const std::size_t copy_end = cached_to - req.to_height;
res.distributions.push_back(
    {amount, req.from_height, vcopy(cached_distribution, copy_start, copy_end), req.to_height}
);
```

Reply:
I assume this must do the same as
res.distributions.push_back({amount, cached_start_height, cached_distribution, cached_base});
but what is the benefit ? Faster again ?

Reply:
The current implementation doesn't use the cache unless the start and end values are exactly the same. It should be possible to use the cache for any subset within the cached range, which is what that code was attempting to do. It will be faster (assuming the full call is really that slow), but will only make a difference if the client is changing the start and end ranges throughout the requests. Perhaps that isn't possible in the wallet implementation currently? It would help still help with any _outside_ implementations using that call (which seem rare ftw).

Reply:
Ah, I see. The wallet should make the call with the same start/stop (though I suppose a public daemon might get different ranges if the wallet has a custom segregation height). I'll change that once I understand what boost adapter does (or make my own copy).

Reply:
Its the same as : 
```c++
std::vector<uint64_t>{
    cached_distribution.begin() + copy_start
  , cached_distribution.begin() + copy_end
}
```
Using boost in this case probably just makes it slightly less readable/clear to shave off a few characters.